awscli/customizations/history/show.py

```diff
@@ -0,0 +1,463 @@
+        formatted_title = self._format_section_title(title, event_record)
+        self._write_output(formatted_title)
+
+    def _display_value(self, value_definition, event_record):
+        value_description = value_definition['description']
+        event_record_payload = event_record['payload']
+        value = event_record_payload
+        if 'payload_key' in value_definition:
+            value = event_record_payload[value_definition['payload_key']]
+        formatted_value = self._format_description(value_description)
+        formatted_value += self._format_value(
+            value, event_record, value_definition.get('value_format')
+        )
+        self._write_output(formatted_value)
+
+    def _write_output(self, content):
+        if isinstance(content, six.text_type):
+            content = content.encode('utf-8')
+        self._output.write(content)
+
+    def _format_section_title(self, title, event_record):
+        formatted_title = title
+        api_num = self._get_api_num(event_record)
+        if api_num is not None:
+            formatted_title = ('[%s] ' % api_num) + formatted_title
+        formatted_title = self._color_if_configured(formatted_title, 'title')
+        formatted_title += '\n'
+
+        formatted_timestamp = self._format_description('at time')
+        formatted_timestamp += self._format_value(
+            event_record['timestamp'], event_record, value_format='timestamp')
+
+        return '\n' + formatted_title + formatted_timestamp
+
+    def _get_api_num(self, event_record):
+        request_id = event_record['request_id']
+        if request_id:
+            if request_id not in self._request_id_to_api_num:
+                self._request_id_to_api_num[
+                    request_id] = self._num_api_calls
+                self._num_api_calls += 1
+            return self._request_id_to_api_num[request_id]
+
+    def _format_description(self, value_description):
+        return self._color_if_configured(
+            value_description + ': ', 'description')
+
+    def _format_value(self, value, event_record, value_format=None):
+        if value_format:
+            formatted_value = getattr(self, '_format_' + value_format)(
+                value, event_record)
+        else:
+            formatted_value = str(value)
+        return formatted_value + '\n'
+
+    def _format_timestamp(self, event_timestamp, event_record=None):
+        return datetime.datetime.fromtimestamp(
+            event_timestamp/1000.0).strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]
+
+    def _format_dictionary(self, obj, event_record=None):
+        return json.dumps(obj=obj, sort_keys=True, indent=4)
+
+    def _format_http_body(self, body, event_record):
+        if not body:
+            return 'There is no associated body'
+        elif event_record['payload'].get('streaming', False):
+            return 'The body is a stream and will not be displayed'
+        elif self._is_xml(body):
```

Comment:
Can you add a TODO here?  From what I can tell, in the happy case where the body is xml, we're parsing the body *three* times (and serializing it twice).  This might be what's causing the large delays/cpu usage I'm seeing for various s3 commands that call `ListObjects`.